// @dojo/webpack-contrib/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
var loader_1 = require("./static-build-loader/loader");
exports.staticOptimizeLoader = loader_1.default;
tslib_1.__exportStar(require("./static-build-loader/loader"), exports);
var loader_2 = require("./css-module-dts-loader/loader");
exports.cssModuleDtsLoader = loader_2.default;
tslib_1.__exportStar(require("./css-module-dts-loader/loader"), exports);
var loader_3 = require("./css-module-decorator-loader/loader");
exports.cssModuleDecoratorLoader = loader_3.default;
tslib_1.__exportStar(require("./css-module-decorator-loader/loader"), exports);
var ExternalLoaderPlugin_1 = require("./external-loader-plugin/ExternalLoaderPlugin");
exports.ExternalLoaderPlugin = ExternalLoaderPlugin_1.default;
tslib_1.__exportStar(require("./external-loader-plugin/ExternalLoaderPlugin"), exports);
var I18nPlugin_1 = require("./i18n-plugin/I18nPlugin");
exports.I18nPlugin = I18nPlugin_1.default;
tslib_1.__exportStar(require("./i18n-plugin/I18nPlugin"), exports);
var BundleAnalyzerPlugin_1 = require("./webpack-bundle-analyzer/BundleAnalyzerPlugin");
exports.BundleAnalyzerPlugin = BundleAnalyzerPlugin_1.default;
tslib_1.__exportStar(require("./webpack-bundle-analyzer/BundleAnalyzerPlugin"), exports);
