// @theintern/leadfoot/keys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var keys = {
    NULL: '\uE000',
    CANCEL: '\uE001',
    HELP: '\uE002',
    BACKSPACE: '\uE003',
    TAB: '\uE004',
    CLEAR: '\uE005',
    RETURN: '\uE006',
    ENTER: '\uE007',
    SHIFT: '\uE008',
    CONTROL: '\uE009',
    ALT: '\uE00A',
    PAUSE: '\uE00B',
    ESCAPE: '\uE00C',
    SPACE: '\uE00D',
    PAGE_UP: '\uE00E',
    PAGE_DOWN: '\uE00F',
    END: '\uE010',
    HOME: '\uE011',
    ARROW_LEFT: '\uE012',
    ARROW_UP: '\uE013',
    ARROW_RIGHT: '\uE014',
    ARROW_DOWN: '\uE015',
    INSERT: '\uE016',
    DELETE: '\uE017',
    SEMICOLON: '\uE018',
    EQUALS: '\uE019',
    NUMPAD0: '\uE01A',
    NUMPAD1: '\uE01B',
    NUMPAD2: '\uE01C',
    NUMPAD3: '\uE01D',
    NUMPAD4: '\uE01E',
    NUMPAD5: '\uE01F',
    NUMPAD6: '\uE020',
    NUMPAD7: '\uE021',
    NUMPAD8: '\uE022',
    NUMPAD9: '\uE023',
    MULTIPLY: '\uE024',
    ADD: '\uE025',
    SEPARATOR: '\uE026',
    SUBTRACT: '\uE027',
    DECIMAL: '\uE028',
    DIVIDE: '\uE029',
    F1: '\uE031',
    F2: '\uE032',
    F3: '\uE033',
    F4: '\uE034',
    F5: '\uE035',
    F6: '\uE036',
    F7: '\uE037',
    F8: '\uE038',
    F9: '\uE039',
    F10: '\uE03A',
    F11: '\uE03B',
    F12: '\uE03C',
    META: '\uE03D',
    COMMAND: '\uE03D',
    ZENKAKU_HANKAKU: '\uE040',
    '\uE000': 'NULL',
    '\uE001': 'Cancel',
    '\uE002': 'Help',
    '\uE003': 'Backspace',
    '\uE004': 'Tab',
    '\uE005': 'Clear',
    '\uE006': 'Return',
    '\uE007': 'Enter',
    '\uE008': 'Shift',
    '\uE009': 'Control',
    '\uE00A': 'Alt',
    '\uE00B': 'Pause',
    '\uE00C': 'Escape',
    '\uE00D': 'Space',
    '\uE00E': 'Page up',
    '\uE00F': 'Page down',
    '\uE010': 'End',
    '\uE011': 'Home',
    '\uE012': 'Left arrow',
    '\uE013': 'Up arrow',
    '\uE014': 'Right arrow',
    '\uE015': 'Down arrow',
    '\uE016': 'Insert',
    '\uE017': 'Delete',
    '\uE018': 'Semicolon',
    '\uE019': 'Equals',
    '\uE01A': 'Numpad 0',
    '\uE01B': 'Numpad 1',
    '\uE01C': 'Numpad 2',
    '\uE01D': 'Numpad 3',
    '\uE01E': 'Numpad 4',
    '\uE01F': 'Numpad 5',
    '\uE020': 'Numpad 6',
    '\uE021': 'Numpad 7',
    '\uE022': 'Numpad 8',
    '\uE023': 'Numpad 9',
    '\uE024': 'Multiply',
    '\uE025': 'Add',
    '\uE026': 'Separator',
    '\uE027': 'Subtract',
    '\uE028': 'Decimal',
    '\uE029': 'Divide',
    '\uE031': 'F1',
    '\uE032': 'F2',
    '\uE033': 'F3',
    '\uE034': 'F4',
    '\uE035': 'F5',
    '\uE036': 'F6',
    '\uE037': 'F7',
    '\uE038': 'F8',
    '\uE039': 'F9',
    '\uE03A': 'F10',
    '\uE03B': 'F11',
    '\uE03C': 'F12',
    '\uE03D': 'Command',
    '\uE040': 'Zenkaku/hankaku'
};
exports.default = keys;
