// @dojo/cli-test-intern/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var chalk_1 = require("chalk");
var path = require("path");
var fs = require("fs");
var runTests_1 = require("./runTests");
var javaCheck_1 = require("./javaCheck");
var pkgDir = require('pkg-dir');
function buildNpmDependencies() {
    try {
        var packagePath = pkgDir.sync(__dirname);
        var packageJsonFilePath = path.join(packagePath, 'package.json');
        var packageJson = require(packageJsonFilePath);
        return packageJson.dependencies;
    }
    catch (e) {
        throw new Error('Failed reading dependencies from package.json - ' + e.message);
    }
}
/**
 * Compiled unit tests should exist when testing against a built application
 */
function assertCompiledTests(args) {
    var projectRoot = pkgDir.sync(process.cwd());
    var unitsPath = path.join(projectRoot, 'output', 'test', 'unit', 'all.js');
    var funcationalsPath = path.join(projectRoot, 'output', 'test', 'functional', 'all.js');
    var hasUnits = args.unit || args.all ? fs.existsSync(unitsPath) : true;
    var hasFunctionals = args.functional || args.all ? fs.existsSync(funcationalsPath) : true;
    if (!hasUnits || !hasFunctionals) {
        throw new Error("Could not find tests" + (args.verbose ? " in " + path.join(projectRoot, 'output', 'test') + ".\nH" : ', h') + "ave you built the tests using dojo build?\n\nFor @dojo/cli-build-app run: dojo build app --mode unit or dojo build app --mode functional");
    }
}
exports.assertCompiledTests = assertCompiledTests;
function transformTestArgs(args) {
    var nodeUnit = true;
    var remoteUnit = false;
    var remoteFunctional = false;
    var internConfig = args.legacy ? 'legacy.json' : 'intern.json';
    if (args.config) {
        assertCompiledTests(args);
    }
    if (args.all) {
        nodeUnit = remoteUnit = remoteFunctional = true;
    }
    if (args.unit) {
        remoteUnit = true;
    }
    if (args.functional) {
        nodeUnit = remoteUnit = false;
        remoteFunctional = true;
    }
    return {
        internConfig: internConfig,
        childConfig: args.config,
        reporters: args.reporters,
        userName: args.userName,
        secret: args.secret,
        testingKey: args.testingKey,
        verbose: args.verbose,
        filter: args.filter,
        externals: args.externals,
        nodeUnit: nodeUnit,
        remoteUnit: remoteUnit,
        remoteFunctional: remoteFunctional
    };
}
function printBrowserLink(options) {
    var browserArgs = [];
    if (options.filter) {
        browserArgs.push('grep=' + encodeURIComponent(options.filter));
    }
    console.log('\n If the project directory is hosted on a local server, unit tests can also be run in browser by navigating to ' +
        chalk_1.default.underline("http://localhost/node_modules/intern/?config=node_modules/@dojo/cli-test-intern/intern/" + options.internConfig + (options.childConfig ? "@" + options.childConfig : '') + (browserArgs.length ? "&" + browserArgs.join('&') : '')));
}
function printGoodbye(options) {
    if (options.childConfig) {
        printBrowserLink(options);
    }
    else {
        printLocalTest();
    }
}
function printLocalTest() {
    console.log('\n These tests were run using Dojo JIT compilation. The test suite may also be run against the built application with ' +
        chalk_1.default.underline('dojo test -c local'));
}
var command = {
    description: 'run unit and/or functional tests for your application',
    register: function (options) {
        options('a', {
            alias: 'all',
            describe: 'Runs unit tests and functional tests. Unit tests are run via node and the local tunnel. Functional tests are run via the local tunnel',
            default: false
        });
        options('c', {
            alias: 'config',
            describe: "Specifies what configuration to test with: 'local'(default), 'headless', 'browserstack', 'testingbot', or 'saucelabs'.",
            type: 'string'
        });
        options('f', {
            alias: 'functional',
            describe: 'Runs only functional tests. Tests are run via the local tunnel',
            default: false
        });
        options('k', {
            alias: 'testingKey',
            describe: 'API key for testingbot or accesskey for saucelabs or browserstack',
            type: 'string'
        });
        options('usr', {
            alias: 'userName',
            describe: 'User name for testing platform',
            type: 'string'
        });
        options('r', {
            alias: 'reporters',
            describe: 'Comma seperated list of reporters [default: "lcov,htmlcoverage,runner"] [ choices: "benchmark", "cobertura", "htmlcoverage", "jsoncoverage", "junit", "lcov", "pretty", "runner", "simple", "teamcity"]',
            type: 'string'
        });
        options('s', {
            alias: 'secret',
            describe: 'API secret for testingbot',
            type: 'string'
        });
        options('u', {
            alias: 'unit',
            describe: 'Runs unit tests via node and the local tunnel',
            default: false
        });
        options('v', {
            alias: 'verbose',
            describe: 'Produce diagnostic messages to the console.',
            default: false
        });
        options('n', {
            alias: 'node',
            describe: 'Run unit tests via node',
            type: 'boolean',
            default: true
        });
        options('l', {
            alias: 'legacy',
            describe: 'Include IE11 when running functional tests',
            type: 'boolean',
            default: false
        });
        options('filter', {
            describe: 'Run only tests whose IDs match a regular expression',
            type: 'string'
        });
    },
    run: function (helper, args) {
        function unhandledRejection(reason) {
            console.log('Unhandled Promise Rejection: ');
            console.log(reason);
        }
        process.on('unhandledRejection', unhandledRejection);
        return javaCheck_1.default(args).then(function (javaCheckPassed) {
            if (javaCheckPassed) {
                var testOptions_1 = transformTestArgs(args);
                return new Promise(function (resolve, reject) {
                    runTests_1.default(testOptions_1)
                        .then(function () {
                        process.removeListener('unhandledRejection', unhandledRejection);
                    })
                        .then(function () {
                        printGoodbye(testOptions_1);
                    }, function (err) {
                        printGoodbye(testOptions_1);
                        throw err;
                    })
                        .then(resolve, reject);
                });
            }
            else {
                return Promise.reject(Error(chalk_1.default.underline('Error! Java VM could not be found.') +
                    '\nA Java VM needs to be installed and available from the command line to allow the ' +
                    chalk_1.default.underline('dojo test') +
                    ' command to run tests in a browser locally or remotely.'));
            }
        });
    },
    eject: function () {
        return {
            npm: {
                devDependencies: tslib_1.__assign({}, buildNpmDependencies())
            },
            copy: {
                path: path.join(__dirname, 'intern'),
                files: ['intern.json', 'legacy.json']
            }
        };
    }
};
exports.default = command;
