// @theintern/digdug/NullTunnel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Task_1 = require("@dojo/core/async/Task");
var Tunnel_1 = require("./Tunnel");
var lang_1 = require("@dojo/core/lang");
var NullTunnel = (function (_super) {
    tslib_1.__extends(NullTunnel, _super);
    function NullTunnel(options) {
        return _super.call(this, lang_1.mixin({
            auth: ''
        }, options || {})) || this;
    }
    Object.defineProperty(NullTunnel.prototype, "isDownloaded", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    NullTunnel.prototype.download = function () {
        return Task_1.default.resolve();
    };
    NullTunnel.prototype.start = function () {
        this._state = 'running';
        return Task_1.default.resolve();
    };
    NullTunnel.prototype.stop = function () {
        this._state = 'stopped';
        return Promise.resolve(0);
    };
    NullTunnel.prototype.sendJobState = function () {
        return Task_1.default.resolve();
    };
    return NullTunnel;
}(Tunnel_1.default));
exports.default = NullTunnel;
