// @dojo/widgets/listbox/ListboxOption.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "@dojo/widget-core/mixins/Themed", "@dojo/widget-core/d", "@dojo/widget-core/WidgetBase", "../theme/listbox.m.css"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var Themed_1 = require("@dojo/widget-core/mixins/Themed");
    var d_1 = require("@dojo/widget-core/d");
    var WidgetBase_1 = require("@dojo/widget-core/WidgetBase");
    var css = require("../theme/listbox.m.css");
    exports.ThemedBase = Themed_1.ThemedMixin(WidgetBase_1.WidgetBase);
    var ListboxOptionBase = /** @class */ (function (_super) {
        tslib_1.__extends(ListboxOptionBase, _super);
        function ListboxOptionBase() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        ListboxOptionBase.prototype._onClick = function (event) {
            event.stopPropagation();
            var _a = this.properties, index = _a.index, key = _a.key, option = _a.option, onClick = _a.onClick;
            onClick && onClick(option, index, key);
        };
        ListboxOptionBase.prototype.render = function () {
            var _a = this.properties, _b = _a.classes, classes = _b === void 0 ? [] : _b, _c = _a.disabled, disabled = _c === void 0 ? false : _c, id = _a.id, label = _a.label, _d = _a.selected, selected = _d === void 0 ? false : _d;
            return d_1.v('div', {
                'aria-disabled': disabled ? 'true' : null,
                'aria-selected': disabled ? null : String(selected),
                classes: this.theme(classes),
                id: id,
                role: 'option',
                onclick: this._onClick
            }, [label]);
        };
        ListboxOptionBase = tslib_1.__decorate([
            Themed_1.theme(css)
        ], ListboxOptionBase);
        return ListboxOptionBase;
    }(exports.ThemedBase));
    exports.ListboxOptionBase = ListboxOptionBase;
    var ListboxOption = /** @class */ (function (_super) {
        tslib_1.__extends(ListboxOption, _super);
        function ListboxOption() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return ListboxOption;
    }(ListboxOptionBase));
    exports.default = ListboxOption;
});
