// @dojo/test-extras/main.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./harness", "./support/assertRender", "./support/callListener", "./support/sendEvent", "./support/d"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var harness_1 = require("./harness");
    exports.harness = harness_1.default;
    var assertRender_1 = require("./support/assertRender");
    exports.assertRender = assertRender_1.default;
    var callListener_1 = require("./support/callListener");
    exports.callListener = callListener_1.default;
    var sendEvent_1 = require("./support/sendEvent");
    exports.sendEvent = sendEvent_1.default;
    var d_1 = require("./support/d");
    exports.assignChildProperties = d_1.assignChildProperties;
    exports.assignProperties = d_1.assignProperties;
    exports.findIndex = d_1.findIndex;
    exports.findKey = d_1.findKey;
    exports.replaceChild = d_1.replaceChild;
    exports.replaceChildProperties = d_1.replaceChildProperties;
    exports.replaceProperties = d_1.replaceProperties;
});
